src/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -172,6 +178,30 @@ internal CallSite<T> CreateMatchMaker()
             return new CallSite<T>();
         }
 
+        internal CallSite GetMatchmaker()
+        {
+            // check if we have a cached matchmaker and attempt to atomically grab it.
+            var matchmaker = _cachedMatchmaker;
+            if (matchmaker != null)
+            {
+                matchmaker = Interlocked.Exchange(ref _cachedMatchmaker, null);
+                Debug.Assert(matchmaker?._match != false, "cached site should be set up for matchmaking");
+            }
+
+            return matchmaker ?? new CallSite<T>() { _match = true };
+        }
+
+        internal void ReleaseMatchmaker(CallSite matchMaker)
+        {
+            // If "Rules" has not been created, this is the first (and likely the only) Update of the site.
+            // 90% sites stay monomorphic and will never need a matchmaker again.
+            // Otherwise store the matchmaker for the future use.
+            if (Rules != null)
+            {
+                _cachedMatchmaker = matchMaker;
```

Comment:
Does this need to be a volatile write?

Reply:
Any use of the object will have to go through`Interlocked.Exchange`
That is a full fence, so regular write is ok here.